content/posts/aws-autoscaling-groups-with-terraform.md

```diff
@@ -0,0 +1,191 @@
+    web_password    = var.web_password,
+    ssl_private_key = tls_private_key.mysvc.private_key_pem,
+    ssl_certificate = tls_self_signed_cert.mysvc.cert_pem
+  }))
+
+  # we don't want to create a new template just because there is a newer AMI
+  lifecycle {
+    ignore_changes = [
+      image_id,
+    ]
+  }
+}
+
+```
+The launch template defines the critical elements of the instance and also loads
+the configuration of the service from the user data file.  The user data file is a template
+that evaluates to a valid cloud-init configuration file.
+
+Now let's create the autoscaling group
+```terraform
+resource "aws_autoscaling_group" "mysvc" {
+  name                      = "mysvc-${aws_launch_template.mysvc.latest_version}"
+  health_check_type         = "ELB"
+  health_check_grace_period = 120
+  termination_policies      = ["OldestInstance"]
+  launch_template {
+    id      = aws_launch_template.mysvc.id
+    version = aws_launch_template.mysvc.latest_version
+  }
+  min_size = 1
+  max_size = 10
+
+  lifecycle {
+    create_before_destroy = true
+  }
+  target_group_arns = [aws_lb_target_group.svc.arn]
+}
+```
+There are a few things to note here:
+
+1. We name the autoscaling group with the latest version of the launch template.  This is to make sure that when the launch template is updated we are creating a new autoscaling group.
+1. When creating a new autoscaling group we want to first create the new one and then destroy the previous one.  Hence the life-cycle directive `create_before_destroy = true`. As our service is behind an ELB this will ensure no down time during the configuration change.
+
+This is great and does exactly what we need.  **Or does it?** 
+
+What happens when we have scaled out the group to have more instances and we change the configuration?
+With the current configuration we will replace the running group of, e.g. 5 instances, with a group of only 1 instance.
+We want to be able to create a new group with 5 instances.
+
+## Better Configuration
+In order to keep the current value of desired capacity we need to somehow get it first.
+
+Let's get the current launch template (if it exists)
+```terraform
+data "aws_launch_template" "current" {
+  filter {
+    name   = "launch-template-name"
+    values = ["mysvc"]
+  }
+}
+```
+
+Now let's see if we have autoscaling groups that use this launch template`s latest version
+```terraform
+data "aws_autoscaling_groups" "current" {
+  filter {
+    name   = "key"
+    values = ["mysvc-template-version"]
+  }
+  filter {
+    name   = "value"
+    values = ["${coalesce(data.aws_launch_template.current.latest_version, 0)}"]
+  }
+}
+```
+> we are using tags that you have not seen being added to the autoscaling group yet.
+> See below for the changes in the autoscaling group resource.
+
+The idea here is that either `data.aws_launch_template.current.latest_version` has a
+value or it is `null`.  If there is a value we expect to have a list of at most size one.
```

Comment:
If there is a value we expect > If there is a value, we expect

Reply:
Fixed